R/expect_type_linter.R

```diff
@@ -0,0 +1,60 @@
+#' Require usage of expect_type(x, type) over expect_equal(typeof(x), type)
+#'
+#' [testthat::expect_type()] exists specifically for testing the storage type
+#'   of objects. [testthat::expect_equal()], [testthat::expect_identical()], and
+#'   [testthat::expect_true()] can also be used for such tests,
+#'   but it is better to use the tailored function instead.
+#'
+#' @evalRd rd_tags("expect_type_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+#' @importFrom glue glue
+expect_type_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    base_type_tests <- xp_text_in_table(paste0("is.", base_types))
+    xpath <- glue::glue("//expr[
+      (
+        SYMBOL_FUNCTION_CALL[text() = 'expect_equal' or text() = 'expect_identical']
+        and following-sibling::expr[1][expr[SYMBOL_FUNCTION_CALL[text() = 'typeof']]]
```

Comment:
Can we use `position() <= 2` here as well?

Reply:
Here I'd prefer not... I can imagine legitimate tests where you're testing the output of a function versus `typeof()`. Something like:

```
expect_equal(GeneratesARawTimestampCheck(x), typeof(Sys.time()))
```

(maybe not the best example, but hopefully conveys my point)

We might prefer `typeof(Sys.time())` instead of `"double"` directly because `"double"` is sort of an implementation detail, and what we really care about is that we match whatever `Sys.time()` is doing, even if that should change.

(maybe a better example would b testing the `typeof()` output of a function in a custom package). 

(maybe a good example is code that produces a factor under R<4 and a string under R>=4, and `typeof()` might provide a concise way of checking our output matches whatever the underlying R version is doing)

Reply:
Good point but doesn't that mean we should check the other argument matches a `STR_CONST` for the lint?

Reply:
I don't think so, `expect_type(expr)` is fine, so `expect_equal(typeof(x), expr)` should still lint.

Reply:
Wait, I'm confused.
What exactly needs linting and what would be a false positive?

`expect_equal(typeof(x), y)`, `expect_equal(y, typeof(x))`, `expect_equal(typeof(x), typeof(y))`, `expect_equal(typeof((x), "constant")`, `expect_equal("constant, typeof((x))`

Reply:
I think all of those _should_ use expect_type, but we are conservative and some would be false negatives.

we could check for string constant in the first argument to cover some more things, WDYT?

Reply:
I see that's what you already suggested. brain fart 🙂

Reply:
But what would a false positive be concretely?
I mean all of the examples could be converted to an equivalent `expect_type()` test, no?

Reply:
in turn:

```
expect_equal(typeof(x), y) # lints already
expect_equal(y, typeof(x)) # could be made to lint
expect_equal(typeof(x), typeof(y)) # lints already
expect_equal(typeof(x), "constant") # lints already
expect_equal("constant", typeof(x)) # lints now
```

Not sure how worth it is to catch the `expect_equal(y, typeof(x))` case. There might be a false positive there I'm not thinking of. e.g. if someone prefers to write tests like `expected <- ...; expect_equal(expected, typeof(x))` and it falls into one of the cases mentioned about where we don't want to lint `typeof()` in the 2nd argument.

I'd prefer to leave it as it is now where we just look for `STR_CONST`.

Reply:
I think we should preserve symmetry WRT the test arguments (i.e., either lint both forms or no forms).
If one form has false positives, so will the other.

Reply:
I don't think the arguments are symmetric, though... the signature is not `x, y` but `object, expected` -- a clear asymmetry in the intended usage.

I'm OK with checking for `STR_CONST` in the first argument because it's a case where we can see a clear violation of the intended usage, without any chance for false positives. So the asymmetry in that case is an exception.

Reply:
Allright. I think I finally understood a false positive could be `expect_equal(my_better_typeof(x), typeof(x))` where `expect_type(x, my_better_typeof(x))` is counter-intuitive.

Reply:
thanks for pressing... agree it's a bit hard to think of false positives for this linter in particular. but this paradigm will also apply to other testthat linters so it's good to address here.